In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("Advertising.csv")

In [ ]:
df

,Unnamed: 0,TV,Radio,Newspaper,Sales
0,1,230.1,37.8,69.2,22.1
1,2,44.5,39.3,45.1,10.4
2,3,17.2,45.9,69.3,9.3
3,4,151.5,41.3,58.5,18.5
4,5,180.8,10.8,58.4,12.9
...,...,...,...,...,...
195,196,38.2,3.7,13.8,7.6
196,197,94.2,4.9,8.1,9.7
197,198,177.0,9.3,6.4,12.8
198,199,283.6,42.0,66.2,25.5


In [ ]:
df = df.drop('Unnamed: 0',axis=1)

In [ ]:
df.head(3)

,TV,Radio,Newspaper,Sales
0,230.1,37.8,69.2,22.1
1,44.5,39.3,45.1,10.4
2,17.2,45.9,69.3,9.3


In [ ]:
from sklearn.model_selection import train_test_split as tts

Extracting features from dataframe to get array or metrics form of data.

In [ ]:
x = df[['TV','Radio','Newspaper']].values
y = df['Sales'].values

In [ ]:
y

array([22.1, 10.4,  9.3, 18.5, 12.9,  7.2, 11.8, 13.2,  4.8, 10.6,  8.6,
       17.4,  9.2,  9.7, 19. , 22.4, 12.5, 24.4, 11.3, 14.6, 18. , 12.5,
        5.6, 15.5,  9.7, 12. , 15. , 15.9, 18.9, 10.5, 21.4, 11.9,  9.6,
       17.4,  9.5, 12.8, 25.4, 14.7, 10.1, 21.5, 16.6, 17.1, 20.7, 12.9,
        8.5, 14.9, 10.6, 23.2, 14.8,  9.7, 11.4, 10.7, 22.6, 21.2, 20.2,
       23.7,  5.5, 13.2, 23.8, 18.4,  8.1, 24.2, 15.7, 14. , 18. ,  9.3,
        9.5, 13.4, 18.9, 22.3, 18.3, 12.4,  8.8, 11. , 17. ,  8.7,  6.9,
       14.2,  5.3, 11. , 11.8, 12.3, 11.3, 13.6, 21.7, 15.2, 12. , 16. ,
       12.9, 16.7, 11.2,  7.3, 19.4, 22.2, 11.5, 16.9, 11.7, 15.5, 25.4,
       17.2, 11.7, 23.8, 14.8, 14.7, 20.7, 19.2,  7.2,  8.7,  5.3, 19.8,
       13.4, 21.8, 14.1, 15.9, 14.6, 12.6, 12.2,  9.4, 15.9,  6.6, 15.5,
        7. , 11.6, 15.2, 19.7, 10.6,  6.6,  8.8, 24.7,  9.7,  1.6, 12.7,
        5.7, 19.6, 10.8, 11.6,  9.5, 20.8,  9.6, 20.7, 10.9, 19.2, 20.1,
       10.4, 11.4, 10.3, 13.2, 25.4, 10.9, 10.1, 16

In [ ]:
y.shape

(200,)

Linear regression math assumes:

𝑦
∈
𝑅
𝑛
×
1
y∈R
n×1

That means:

 y
=
[
22.1


10.4


9.3
]


y=
	​
[
22.1
10.4
9.3
]

But NumPy’s (n,) array is neither:

(n,1) → column vector ❌

(1,n) → row vector ❌

So when you do matrix math like:

means:

“I want 1 column, and NumPy should figure out how many rows are needed.”

The -1 is NumPy’s way of saying:

“Compute this automatically.”

In [ ]:
y = y.reshape(-1,1)
y.shape

(200, 1)

here we get our shape as n*1

In [ ]:
x.shape

(200, 3)

Adding bias intercept column

In [ ]:
x.shape[0]

200

In [ ]:
m = x.shape[0]
x_b = np.hstack([np.ones((m,1)),x])

In [ ]:
x_b

array([[  1. , 230.1,  37.8,  69.2],
       [  1. ,  44.5,  39.3,  45.1],
       [  1. ,  17.2,  45.9,  69.3],
       [  1. , 151.5,  41.3,  58.5],
       [  1. , 180.8,  10.8,  58.4],
       [  1. ,   8.7,  48.9,  75. ],
       [  1. ,  57.5,  32.8,  23.5],
       [  1. , 120.2,  19.6,  11.6],
       [  1. ,   8.6,   2.1,   1. ],
       [  1. , 199.8,   2.6,  21.2],
       [  1. ,  66.1,   5.8,  24.2],
       [  1. , 214.7,  24. ,   4. ],
       [  1. ,  23.8,  35.1,  65.9],
       [  1. ,  97.5,   7.6,   7.2],
       [  1. , 204.1,  32.9,  46. ],
       [  1. , 195.4,  47.7,  52.9],
       [  1. ,  67.8,  36.6, 114. ],
       [  1. , 281.4,  39.6,  55.8],
       [  1. ,  69.2,  20.5,  18.3],
       [  1. , 147.3,  23.9,  19.1],
       [  1. , 218.4,  27.7,  53.4],
       [  1. , 237.4,   5.1,  23.5],
       [  1. ,  13.2,  15.9,  49.6],
       [  1. , 228.3,  16.9,  26.2],
       [  1. ,  62.3,  12.6,  18.3],
       [  1. , 262.9,   3.5,  19.5],
       [  1. , 142.9,  29.3,  12.6],
 

here we get our shape as n*(m+1)

Transposing x

In [ ]:
x_t = x_b.T

In [ ]:
x_t

array([[  1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
          1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,   1. ,
      

What does @ mean?

Matrix multiplication (not element-wise)

Mathematically:

𝑋
𝑇
𝑋


This matrix:

captures correlations between features

must be invertible for the normal equation

In [ ]:
xtx = x_t @ x_b

In [ ]:
xtx

array([[2.00000000e+02, 2.94085000e+04, 4.65280000e+03, 6.11080000e+03],
       [2.94085000e+04, 5.79111839e+06, 6.98061980e+05, 9.19625280e+05],
       [4.65280000e+03, 6.98061980e+05, 1.52107860e+05, 1.64946550e+05],
       [6.11080000e+03, 9.19625280e+05, 1.64946550e+05, 2.81096740e+05]])

Inverse of xtx

In [ ]:
xtx_inv = np.linalg.inv(xtx)

In [ ]:
xtx_inv

array([[ 3.42444998e-02, -9.35348333e-05, -3.92647297e-04,
        -2.08036831e-04],
       [-9.35348333e-05,  6.84890750e-07, -1.57355890e-07,
        -1.14959987e-07],
       [-3.92647297e-04, -1.57355890e-07,  2.61016474e-05,
        -6.26574016e-06],
       [-2.08036831e-04, -1.14959987e-07, -6.26574016e-06,
         1.21328472e-05]])

xt*y

In [ ]:
xt_y = x_t @ y

Final beta computation

In [ ]:
beta = xtx_inv @ xt_y

In [ ]:
beta

array([[ 2.93888937e+00],
       [ 4.57646455e-02],
       [ 1.88530017e-01],
       [-1.03749304e-03]])

In [ ]:
print("Intercept:", beta[0])
print("TV coefficient:", beta[1])
print("Radio coefficient:", beta[2])
print("Newspaper coefficient:", beta[3])

Intercept: [2.93888937]
TV coefficient: [0.04576465]
Radio coefficient: [0.18853002]
Newspaper coefficient: [-0.00103749]


In [ ]:
y_pred = x_b @ beta

In [ ]:
y_pred

array([[20.52397441],
       [12.33785482],
       [12.30767078],
       [17.59782951],
       [13.18867186],
       [12.47834763],
       [11.72975995],
       [12.12295317],
       [ 3.72734086],
       [12.55084872],
       [ 7.0322992 ],
       [17.28512918],
       [10.57712073],
       [ 8.82630048],
       [18.43436638],
       [20.81929952],
       [12.82365674],
       [23.22495716],
       [ 9.95168206],
       [14.16607293],
       [18.10076728],
       [14.7405382 ],
       [ 6.4891503 ],
       [16.5459329 ],
       [ 8.14651887],
       [15.6100386 ],
       [14.98951429],
       [17.05167344],
       [19.41053803],
       [ 9.14402389],
       [21.6339338 ],
       [11.3460929 ],
       [ 7.63888314],
       [18.86426829],
       [ 7.57483051],
       [17.00682618],
       [23.40590052],
       [15.62347779],
       [ 9.90868103],
       [20.44761039],
       [16.37766467],
       [17.2959832 ],
       [21.59580326],
       [13.96385684],
       [ 8.88787996],
       [15

<h1> modular coding

In [ ]:
# x_train, x_test, y_train, y_test = tts(x,y,test_size=0.2,random_state=42)

In [ ]:
# print(x_train.shape)
# print(y_train.shape)
# print(x_test.shape)
# print(y_test.shape)

In [ ]:
# x_train

<h1> using normal equation on train test split

In [ ]:
def multi_lr(x: pd.DataFrame, y: pd.Series):

    x_train, x_test, y_train, y_test = tts(
        x, y, test_size=0.2, random_state=42
    )

    # Ensure y_train is column vector
    y_train = np.asarray(y_train).reshape(-1, 1)

    # Convert X to NumPy
    x_train = np.asarray(x_train)
    x_test = np.asarray(x_test)

    # Add bias column
    m = x_train.shape[0]
    x_b = np.hstack([np.ones((m, 1)), x_train])

    # Normal Equation
    x_t = x_b.T
    beta = np.linalg.inv(x_t @ x_b) @ (x_t @ y_train)

    print("Intercept:", beta[0])
    print("TV coefficient:", beta[1])
    print("Radio coefficient:", beta[2])
    print("Newspaper coefficient:", beta[3])

    # Prediction
    m_test = x_test.shape[0]
    x_test_b = np.hstack([np.ones((m_test, 1)), x_test])
    y_pred = x_test_b @ beta

    return beta, y_pred, x_test, y_test

In [ ]:
lr = multi_lr(x,y)

Intercept: [2.97906734]
TV coefficient: [0.04472952]
Radio coefficient: [0.18919505]
Newspaper coefficient: [0.00276111]


In [ ]:
from sklearn.metrics import r2_score

# Unpack x_test, y_pred, and y_test from the lr tuple
_, y_pred, x_test, y_test = lr

r2_score(y_test, y_pred)

0.8994380241009124

<h1> using Normal equation full dataset training

In [ ]:
def multi_lr(x: pd.Series, y: pd.Series):

  # reshape y to n*1
  y = y.reshape(-1,1)
  y.shape

  # adding bias column into x
  m = x.shape[0]
  x_b = np.hstack([np.ones((m,1)),x])

  # transposing of x_b
  x_t = x_b.T

  # x_t * x_b
  xtx = x_t @ x_b

  # inverse of xtx
  xtx_inv = np.linalg.inv(xtx)

  # x_t * y
  xt_y = x_t @ y

  # final beta computation
  beta = xtx_inv @ xt_y

  print("Intercept:", beta[0])
  print("TV coefficient:", beta[1])
  print("Radio coefficient:", beta[2])
  print("Newspaper coefficient:", beta[3])

  y_pred = x_b @ beta

  return beta, y_pred


In [ ]:
ml = multi_lr(x,y)

Intercept: [2.93888937]
TV coefficient: [0.04576465]
Radio coefficient: [0.18853002]
Newspaper coefficient: [-0.00103749]


In [ ]:
predictions = ml[1]

In [ ]:
predictions

array([[20.52397441],
       [12.33785482],
       [12.30767078],
       [17.59782951],
       [13.18867186],
       [12.47834763],
       [11.72975995],
       [12.12295317],
       [ 3.72734086],
       [12.55084872],
       [ 7.0322992 ],
       [17.28512918],
       [10.57712073],
       [ 8.82630048],
       [18.43436638],
       [20.81929952],
       [12.82365674],
       [23.22495716],
       [ 9.95168206],
       [14.16607293],
       [18.10076728],
       [14.7405382 ],
       [ 6.4891503 ],
       [16.5459329 ],
       [ 8.14651887],
       [15.6100386 ],
       [14.98951429],
       [17.05167344],
       [19.41053803],
       [ 9.14402389],
       [21.6339338 ],
       [11.3460929 ],
       [ 7.63888314],
       [18.86426829],
       [ 7.57483051],
       [17.00682618],
       [23.40590052],
       [15.62347779],
       [ 9.90868103],
       [20.44761039],
       [16.37766467],
       [17.2959832 ],
       [21.59580326],
       [13.96385684],
       [ 8.88787996],
       [15

<h1> using class

In [ ]:
x_train, x_test, y_train, y_test = tts(x,y,test_size=0.2,random_state=42)

In [ ]:
y_train.shape

(160, 1)

In [ ]:
class mlr:

  def __init__(self):
    self.coeff_ = None
    self.Intercept = None

  def train(self, x_train, y_train):
    x_train = np.insert(x_train, 0, 1, axis=1)

    beta = np.linalg.inv(np.dot(x_train.T, x_train)).dot(x_train.T).dot(y_train)

    self.coeff_ = beta[1:]
    self.Intercept = beta[0]

  def predict(self,x_test):
    y_pred = np.dot(x_test,self.coeff_) + self.Intercept
    return y_pred

In [ ]:
lr = mlr()

In [ ]:
lr.train(x_train,y_train)

In [ ]:
prediction = lr.predict(x_test)

In [ ]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test,prediction)

In [ ]:
r2

0.8994380241009123

both tts and class model get the same accuracy

<h1> linear reg class

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
x = df.drop('Sales',axis=1)
y = df['Sales']

x_train,x_test,y_train,y_test = tts(x,y,test_size=0.2,random_state=42)

In [ ]:
lr = LinearRegression()

lr.fit(x_train,y_train)
pre = lr.predict(x_test)

In [ ]:
r2_score(y_test,pre)

0.899438024100912